language/stdlib/modules/Roles.move

```diff
@@ -381,61 +380,238 @@ module Roles {
+        apply ValidatorOperatorRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    spec schema ParentVASPRoleMatchesRoleId {
+        invariant module forall addr: address where exists<Privilege<ParentVASPRole>>(addr):
+             spec_has_role_id(addr)
+                && (spec_get_role_id(addr) == SPEC_PARENT_VASP_ROLE_ID());
+
+        invariant module forall addr: address where spec_has_role_id(addr):
+            (spec_get_role_id(addr) == SPEC_PARENT_VASP_ROLE_ID())
+                ==> exists<Privilege<ParentVASPRole>>(addr);
+    }
+    spec module {
+        apply ParentVASPRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    spec schema ChildVASPRoleMatchesRoleId {
+        invariant module forall addr: address where exists<Privilege<ChildVASPRole>>(addr):
+             spec_has_role_id(addr)
+                && (spec_get_role_id(addr) == SPEC_CHILD_VASP_ROLE_ID());
+
+        invariant module forall addr: address where spec_has_role_id(addr):
+            (spec_get_role_id(addr) == SPEC_CHILD_VASP_ROLE_ID())
+                ==> exists<Privilege<ChildVASPRole>>(addr);
+    }
+    spec module {
+        apply ChildVASPRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    spec schema UnhostedRoleMatchesRoleId {
+        invariant module forall addr: address where exists<Privilege<UnhostedRole>>(addr):
+             spec_has_role_id(addr)
+                && (spec_get_role_id(addr) == SPEC_UNHOSTED_ROLE_ID());
+
+        invariant module forall addr: address where spec_has_role_id(addr):
+            (spec_get_role_id(addr) == SPEC_UNHOSTED_ROLE_ID())
+                ==> exists<Privilege<UnhostedRole>>(addr);
+    }
+    spec module {
+        apply UnhostedRoleMatchesRoleId to public *, public *<T>;
+    }
+
+    /// ## Role persistence
+
+    /// **Informally:** Once an account at address `A` is granted a role `R` it
+    /// will remain an account with role `R` for all time.
+    spec schema RoleIdPersists {
+        ensures forall addr: address where old(spec_has_role_id(addr)) :
+            spec_has_role_id(addr) && (old(spec_get_role_id(addr)) == spec_get_role_id(addr));
+    }
+
+    spec module {
+        apply RoleIdPersists to *<T>, *;
+    }
+
+    /// ## Privilege extraction
+
+/// >TODO: This has not been proved yet and may be wrong, or need further
+/// strengthening.  Or, it may just be too hard for the prover.
+/// The capability returned from extract_privilege_to_capability is free-floating,
+/// so we have to quantify over members of the type, not over addresses or something.
+
+    spec schema IsExtractedIffCapabilityExists {
+        /// **Informally:** No capabilities for a privilege witness at a given
+        /// address exist until the privilege exists.
+        /// > NOTE: This verifies by itself when applied to everything.
+        invariant forall priv: type, addr: address where !exists<Privilege<priv>>(addr):
+            !(exists c: Capability<priv>: c.owner_address == addr);
```

Comment:
I think this doesn't make sense to me. You can indeed construct a `c` for which this does not hold. The particular function which does the extract should have this as a post condition. Why make this even a global spec? 

Also you don't need to quantify priv, you can use it as a type parameter to the schema, because only generic functions can even talk about Privilege<priv>. 

I'm fine we submit this for now as an experiment but I think this should be done differently.

Reply:
1.  !exists ... says there does not exist a capability c for that privilege and address, which is what I wanted (I think).
2. I think it needs to be an invariant, because it's going to be true for an arbitrarily long time until someone registers the privilege at the address.  It's forall time/ forall state, so we need to show that all function calls preserve it.

I *think* it's necessary to prove the other (commented out) properties here, but I was working fast and late at night.